### PR TITLE
router: update all http filters to get per filter config by the http filter config name at first

### LINKED ISSUE #12274: allow typed_per_filter_config to refer to "custom" HTTP filter names
Use case: We'd like to use multiple ext_authz filters configured with different gRPC servers, and would like to be able to toggle them separately via route config. This means we can't just use the canonical filter name, but need to be able to refer to custom names.

https://github.com/envoyproxy/envoy/pull/9618 made Envoy able to infer most extension types from the type URL of the config, which frees up users to use the name field to make filter names more descriptive (particularly useful for distinguishing multiple instances of the same filter).

It looks like typed_per_filter_config doesn't yet have this capability and requires that the canonical filter name is used.

I'm not super familiar with the code or filter registration system, but it seems like this could be done by either:
1) keeping a map of filter name to type, and looking up the supplied name in that map to figure out the filter config type
2) use the type URL of the config to infer the filter type
3) both (with a precedence order for applying 1 and 2).

Probably just 1) makes sense? that would ensure that there can't be multiple configs that map to the same filter.

It also might require a config option somewhere backwards compatible? There could be a working config today where a filter has a custom name but the typed_per_filter_config is using the old name.

### PR DESCRIPTION
**Pull request by wbpcode**
Commit Message: router: update all http filters to get per filter config by the http filter config name

Additional Description:

Further work of #21525. Last step to close #12274. 

Risk Level: Mid.
Testing: Updated.
Docs Changes: n/a.
Release Notes: Added.
Platform Specific Features: n/a.

### CODE REVIEW

**Comment by wbpcode** [comment]
@soulxu could you take a look as a first pass when you have free time? Thanks.

**Review by soulxu — COMMENTED** [review]
LGTM except one comment about duplicated test.

File: test/common/http/utility_test.cc [hunk 1]
```diff
@@ -1068,35 +1068,34 @@ class TestConfig : public Router::RouteSpecificFilterConfig {
 
 // Verify that it resolveMostSpecificPerFilterConfig works with nil routes.
 TEST(HttpUtility, ResolveMostSpecificPerFilterConfigNilRoute) {
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply test/common/http/utility_test.cc:1070]: get it. done.

**Comment by wbpcode** [comment]
cc @markdroth could you take a look when you have free time for this doc only API changes? Thanks.

**Comment by wbpcode** [comment]
cc @kyessenov @mattklein123 I think this is also ready for a second pass. Could you take a look? Thanks.

File: api/envoy/config/route/v3/route_components.proto [hunk 2]
```diff
@@ -145,11 +145,15 @@ message VirtualHost {
...
   CorsPolicy cors = 8;
 
-  // The per_filter_config field can be used to provide virtual host-specific
-  // configurations for filters. The key should match the filter name, such as
-  // *envoy.filters.http.buffer* for the HTTP buffer filter. Use of this field is filter
-  // specific; see the :ref:`HTTP filter documentation <config_http_filters>`
-  // for if and how it is utilized.
+  // The per_filter_config field can be used to provide virtual host-specific configurations for filters.
+  // The key should match the `filter config name
+  // <envoy_v3_api_field_extensions.filters.network.http_connection_manager.v3.HttpFilter.name>`.
```
[the remark this answers is not in the document]
Comment (wbpcode) [reply api/envoy/config/route/v3/route_components.proto:150]: Get it, thanks.

**Review by mattklein123 — APPROVED** [review]
Thanks!

**Pull request merged**